### ISSUE TITLE
optionURLReflectsImperativeAPIs not working with ShellRoutes

### ISSUE THREAD

**Issue by JonasJW** [flutter/flutter#146054]
### Steps to reproduce

1. Use the code reproducible code to setup the `demoShellRouter` in a MaterialApp
2. Run the code on web
3. Click the `Post 1` button
4. Notice the URL does not change from `http://localhost:5000/posts`
5. Now use the `demoRouter` in the MaterialApp
6. Run the code on web
7. Click the `Post 1` button
8. Notice the URL changes to `http://localhost:5000/posts/1`

### Expected results

When routing with `GoRouter.of(...).push`, it should update the URL in the browser when setting `GoRouter.optionURLReflectsImperativeAPIs = true;`

### Actual results

`GoRouter.of(...).push` does not update the URL of the browser when setting `GoRouter.optionURLReflectsImperativeAPIs = true;` and using `ShellRoute`.

### Code sample

<details open><summary>GoRouter definitions</summary>

```dart
final demoRouter = GoRouter(
  initialLocation: "/posts",
  routes: [
    GoRoute(
      path: "/posts",
      builder: (context, state) {
        return PostsPage();
      },
      routes: [
        GoRoute(
          path: ":id",
          builder: (context, state) {
            return PostPage(id: state.pathParameters["id"] ?? "0");
          },
        ),
      ],
    )
  ],
);

final demoShellRouter = GoRouter(
  initialLocation: "/posts",
  routes: [
    ShellRoute(
      builder: (context, state, child) {
        return HomePage(child: child);
      },
      routes: [
        GoRoute(
          path: "/posts",
          builder: (context, state) {
            return PostsPage();
          },
          routes: [
            GoRoute(
              path: ":id",
              builder: (context, state) {
                return PostPage(id: state.pathParameters["id"] ?? "0");
              },
            ),
          ],
        )
      ],
    ),
  ],
);


class HomePage extends StatelessWidget {

  final Widget child;

  const HomePage({super.key, required this.child});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Home"),
      ),
      body: child,
    );
  }
}

class PostsPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Posts"),
      ),
      body: TextButton(
        onPressed: () {
          context.push("/posts/1");
        },
        child: Text("Go to post 1"),
      ),
    );
  }
}

class PostPage extends StatelessWidget {
  final String id;

  const PostPage({super.key, required this.id});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Post $id"),
      ),
      body: Text("Post $id"),
    );
  }
}
```
</details>

<details open><summary>MaterialApp definition</summary>

```dart
@override
Widget build(BuildContext context) {
  GoRouter.optionURLReflectsImperativeAPIs = true;

  return MaterialApp.router(
    title: "Test App",
    locale: DevicePreview.locale(context),
    debugShowCheckedModeBanner: false,
    builder: (context, child) {
      return child!;
    },
    routeInformationProvider: demoRouter.routeInformationProvider, // use demoRouter and/or demoShellRouter
    routeInformationParser: demoRouter.routeInformationParser,
    routerDelegate: demoRouter.routerDelegate,
  );
}
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.4, on macOS 14.4 23E214 darwin-arm64, locale en-DE)
    • Flutter version 3.19.4 on channel stable at /Users/jonas/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 68bfaea224 (11 days ago), 2024-03-20 15:36:31 -0700
    • Engine revision a5c24f538d
    • Dart version 3.3.2
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/jonas/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.87.2)
    • VS Code at /Users/jonas/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.85.20240301

[✓] Connected device (3 available)
    • iPhone 15 Pro Max (mobile) • B95DD896-131D-4163-A83C-1AC8A70E8560 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-17-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 14.4 23E214 darwin-arm64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 123.0.6312.87

[✓] Network resources
    • All expected network resources are available.

```

</details>

**Comment by Levi-Lesches** [comment]
I'm experiencing the same issue, using a StatefulShellRoute

**Comment by migueldeoleiros** [comment]
This is a duplicate of #142053
It was fixed on go_router 13.2.2

**Comment by JonasJW** [comment]
@migueldeoleiros you are right, it works on `13.2.2`. Thank you!
